docs/apps/rbac.md

```diff
@@ -0,0 +1,538 @@
+model of `MyModel`, or `None`.
+
+TODO: Update to allow ManyToMany parent relationship in addition to ForeignKey
+https://github.com/ansible/django-ansible-base/issues/78
+
+#### Django Model, Apps, and Permission Constraints
+
+It is fine to register models from multiple apps, but among the registered models,
+the `._meta.model_name` must be unique. If, in the above example,
+you registered `my_app.mymodel` then it is fine if other models by that same
+name exist like `your_app.mymodel`, but _registering both_ models will throw an error.
+
+A similar constraint exists for the permission `codename`.
+If the `Meta` of `MyModel` lists `permissions = [("move", "Can move my model")]`,
+then no other _registered_ model should list that permission.
+
+This rules out things like registering a model, and also registering a proxy
+model _of that model_.
+
+### Parent Resources
+
+Assuming `obj` has a related `organization` which was declared by `parent_field_name` when registering,
+you can give and remove permissions to all objects of that type "in" the organization
+(meaning they have that ForeignKey to the organization).
+- `rd.give_permission(user, organization)` - give execute/view permissions to all job templates in that organization
+- `rd.remove_permission(user, organization)` - revoke permissions obtained from that particular role (other roles will still be in effect)
+
+### Evaluating Permissions
+
+The ultimate goal of this system is to evaluate what objects a user
+has a certain type of permission to (access control).
+Given a registered model, you can do these things.
+
+- obtain visible objects for a user with `MyModel.access_qs(user, 'view_mymodel')`
+- obtain objects user has permission to delete `MyModel.access_qs(user, 'delete_mymodel')`
+- determine if user can delete one specific object `user.has_obj_perm(obj, 'delete_mymodel')`
+- use only the action name for a queryset shortcut `MyModel.access_qs(user, 'view')`
+- get visible objects, view permission implied `MyModel.access_qs(user)`
+- use only the action name or object permission check `user.has_obj_perm(obj, 'delete')`
+- efficient filtering of related model `RelatedModel.objects.filter(mymodel=MyModel.access_ids_qs(user))`
+
+Some HTTP actions will be more complicated. For instance, if you create a new object that combines
+several related objects and each of those related objects require "use" permission.
+Those cases are expected to make multiple calls to methods like `has_obj_perm` within the
+API code, including views, permission classes, serializer classes, templates, forms, etc.
+
+### Creator Permissions
+
+You can give a user "add" permission to a parent model, like "add_mymodel".
+The Django `Permission` entry links to the content type for `MyModel`,
+but these permissions can only apply to the parent model of the object,
+or act as system-wide permissions.
+After a user creates an object, you need to give the user creator permissions.
```

Comment:
> It seems like that second step should be an edit or something on the child.

Yes, it is this. It gives `settings.ANSIBLE_BASE_CREATOR_DEFAULTS` permissions. _Presumably_ you include "change" in that list. Wording is a little tricky here, because of course, it has to have a role definition to do this. So it goes through kind of a wacky complicated process. Firstly it checks if you have all the permissions to the object defined in that setting, so if you do, nothing new is assigned. This should happen quite commonly. Next, it looks to see if an existing role definition gives exactly the requested permissions, and if so, it gives you that role. Finally, the awkward case can happen - there's no existing role to give you those permissions. So it creates a new role and gives you that. This is the challenge in supporting the "add" permission, and there is some precedent of this happening in AWX, although AWX didn't have custom roles, leading to these extra steps...